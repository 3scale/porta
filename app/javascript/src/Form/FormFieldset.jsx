// @flow
// TODO: Replace this component when patternfly-react implements it.

import * as React from 'react'
// $FlowFixMe Flow has troubles with @patternfly modules
import { FormContext } from '@patternfly/react-core/dist/js/components/Form/FormContext'
// $FlowFixMe Flow has troubles with @patternfly modules
import styles from '@patternfly/react-styles/css/components/Form/form'
// $FlowFixMe Flow has troubles with @patternfly modules
import { css } from '@patternfly/react-styles'

type Props = {
  children?: React.Node,
  className?: string,
  isInline?: boolean
}

const FormFieldset = ({
  children,
  className = '',
  isInline = false,
  ...props
}: Props): React.Node => (
  <FormContext.Consumer>
    {({ isHorizontal }: { isHorizontal: boolean }) => (
      <fieldset
        {...props}
        className={css(styles.formFieldset, isInline ? styles.modifiers.inline : className)}
      >
        {isHorizontal ? <div className={css(styles.formHorizontalGroup)}>{children}</div> : children}
      </fieldset>
    )}
  </FormContext.Consumer>
)

export {
  FormFieldset
}
