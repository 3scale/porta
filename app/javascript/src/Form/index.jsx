// @flow

export * from 'Form/FormFieldset'
export * from 'Form/FormLegend'
