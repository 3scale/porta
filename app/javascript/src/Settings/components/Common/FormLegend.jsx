// @flow
// TODO: Replace this component when patternfly-react implements it.

import * as React from 'react'

type Props = {
  children?: React.Node,
  className?: string
}

const FormLegend = ({
  children,
  className = '',
  ...props
}: Props): React.Node => (
  <legend {...props} className={`pf-c-form__legend ${className}`}>
    {children}
  </legend>
)

export {
  FormLegend
}
