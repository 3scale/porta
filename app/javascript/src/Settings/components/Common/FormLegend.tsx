// TODO: Replace this component when patternfly-react implements it.

import * as React from 'react'

type Props = {
  children?: React.ReactNode,
  className?: string
};

const FormLegend = (
  {
    children,
    className = '',
    ...props
  }: Props
): React.ReactElement => <legend {...props} className={`pf-c-form__legend ${className}`}>
  {children}
</legend>

export {
  FormLegend
}
