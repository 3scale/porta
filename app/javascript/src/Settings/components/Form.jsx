// @flow

import * as React from 'react'
import { useState } from 'react'
import { FormFieldset, FormLegend } from 'Form'
import { FormCollection, TextInputGroup, RadioFieldset } from 'Settings/components/Common'
import { AuthenticationSettingsFieldset } from 'Settings/components/AuthenticationSettingsFieldset'
import type { SettingsProps as Props } from 'Settings/types'

const SERVICE_MESH_INTEGRATION = 'service_mesh_istio'
const PROXY_HOSTED_INTEGRATION = 'hosted'

const Form = ({
  isProxyCustomUrlActive,
  integrationMethod,
  authenticationMethod,
  proxyEndpoints,
  authenticationSettings,
  credentialsLocation,
  security,
  gatewayResponse
}: Props): React.Node => {
  const [ selectedIntegrationMethod, setSelectedIntegrationMethod ] = useState(integrationMethod.value)
  const [ selectedAuthenticationMethod, setSelectedAuthenticationMethod ] = useState(authenticationMethod.value)
  const onChange = (setState) => (_checked, e) => setState(e.currentTarget.value)
  const isServiceMesh = selectedIntegrationMethod === SERVICE_MESH_INTEGRATION
  const isProxyHosted = selectedIntegrationMethod === PROXY_HOSTED_INTEGRATION
  const isProxyUrlsReadOnly = !isProxyCustomUrlActive && isProxyHosted
  const customProxyEndpoints = proxyEndpoints.map(endpoint =>
    ({ ...endpoint, readOnly: isProxyUrlsReadOnly, isDefaultValue: isProxyUrlsReadOnly }))

  return (
    <React.Fragment>
      <RadioFieldset {...integrationMethod} onChange={onChange(setSelectedIntegrationMethod)} value={selectedIntegrationMethod} legend='Integration' />
      { !isServiceMesh && <FormCollection collection={customProxyEndpoints} ItemComponent={TextInputGroup} legend='API gateway' /> }
      <RadioFieldset {...authenticationMethod} onChange={onChange(setSelectedAuthenticationMethod)} value={selectedAuthenticationMethod} legend='Authentication' />
      {/* $FlowIgnore[cannot-spread-inexact] */}
      <AuthenticationSettingsFieldset
        isServiceMesh={isServiceMesh}
        authenticationMethod={selectedAuthenticationMethod}
        {...authenticationSettings}
      />
      { !isServiceMesh && <React.Fragment>
        <RadioFieldset {...credentialsLocation} legend='Credentials Location' />
        <FormCollection collection={security} ItemComponent={TextInputGroup} legend='Security' />
        <FormFieldset id='fieldset-GatewayResponse'>
          <FormLegend>Gateway Response</FormLegend>
          {gatewayResponse.map(settings => (
            <FormCollection key={settings.legend} collection={settings.collection} ItemComponent={TextInputGroup} legend={settings.legend} />
          ))}
        </FormFieldset>
      </React.Fragment> }
    </React.Fragment>
  )
}

export {
  Form
}
