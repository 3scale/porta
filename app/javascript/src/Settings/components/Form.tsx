import { useState } from 'react'

import {
  FormCollection,
  FormFieldset,
  FormLegend,
  TextInputGroup,
  RadioFieldset
} from 'Settings/components/Common'
import { AuthenticationSettingsFieldset } from 'Settings/components/AuthenticationSettingsFieldset'
import { createReactWrapper } from 'utilities/createReactWrapper'
import { SETTINGS_DEFAULT } from 'Settings/defaults'
import { FieldCatalogProps, FieldGroupProps, LegendCollectionProps, TypeItemProps } from 'Settings/types'

const SERVICE_MESH_INTEGRATION = 'service_mesh_istio'
const PROXY_HOSTED_INTEGRATION = 'hosted'

type Props = {
  isProxyCustomUrlActive: boolean,
  integrationMethod: FieldGroupProps & FieldCatalogProps,
  authenticationMethod: FieldGroupProps & FieldCatalogProps,
  proxyEndpoints: FieldGroupProps[],
  authenticationSettings: {
    appIdKeyPairSettings: FieldGroupProps[],
    apiKeySettings: FieldGroupProps,
    oidcSettings: {
      basicSettings: TypeItemProps,
      flowSettings: FieldGroupProps[],
      jwtSettings: TypeItemProps
    }
  },
  credentialsLocation: FieldGroupProps & FieldCatalogProps,
  security: FieldGroupProps[],
  gatewayResponse: LegendCollectionProps[]
}

const Form: React.FunctionComponent<Props> = ({
  isProxyCustomUrlActive,
  integrationMethod,
  authenticationMethod,
  proxyEndpoints,
  authenticationSettings,
  credentialsLocation,
  security,
  gatewayResponse
}) => {
  const [ selectedIntegrationMethod, setSelectedIntegrationMethod ] = useState(integrationMethod.value)
  const [ selectedAuthenticationMethod, setSelectedAuthenticationMethod ] = useState(authenticationMethod.value)
  const onChange = (setState: (arg1: ((arg1: string) => string) | string) => void) => (_checked: string, e: React.SyntheticEvent<HTMLButtonElement>) => setState(e.currentTarget.value)
  const isServiceMesh = selectedIntegrationMethod === SERVICE_MESH_INTEGRATION
  const isProxyHosted = selectedIntegrationMethod === PROXY_HOSTED_INTEGRATION
  const isProxyUrlsReadOnly = !isProxyCustomUrlActive && isProxyHosted
  const customProxyEndpoints = proxyEndpoints.map(endpoint => ({
    ...endpoint,
    readOnly: isProxyUrlsReadOnly,
    isDefaultValue: isProxyUrlsReadOnly
  }))

  return (
     <>
      <RadioFieldset {...integrationMethod} onChange={onChange(setSelectedIntegrationMethod)} value={selectedIntegrationMethod} legend='Integration' />
      { !isServiceMesh && <FormCollection collection={customProxyEndpoints} ItemComponent={TextInputGroup} legend='API gateway' /> }
      <RadioFieldset {...authenticationMethod} onChange={onChange(setSelectedAuthenticationMethod)} value={selectedAuthenticationMethod} legend='Authentication' />
      {/* $FlowIgnore[cannot-spread-inexact] */}
      <AuthenticationSettingsFieldset
        isServiceMesh={isServiceMesh}
        authenticationMethod={selectedAuthenticationMethod}
        {...authenticationSettings}
      />
      { !isServiceMesh &&  <>
        <RadioFieldset {...credentialsLocation} legend='Credentials Location' />
        <FormCollection collection={security} ItemComponent={TextInputGroup} legend='Security' />
        <FormFieldset id='fieldset-GatewayResponse'>
          <FormLegend>Gateway Response</FormLegend>
          {gatewayResponse.map(settings => (
            <FormCollection key={settings.legend} collection={settings.collection} ItemComponent={TextInputGroup} legend={settings.legend} />
          ))}
        </FormFieldset>
       </> }
     </>
  )
}

const FormWrapper = (settings: Props = SETTINGS_DEFAULT, elementId: string) => createReactWrapper(<Form {...settings} />, elementId)

export { Form, FormWrapper, Props }
