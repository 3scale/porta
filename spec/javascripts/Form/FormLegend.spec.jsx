// @flow

import React from 'react'
import { shallow } from 'enzyme'
import { FormLegend } from 'Form/FormLegend'

test('FormLegend', () => {
  const view = shallow(<FormLegend>I am Legendary</FormLegend>)
  expect(view).toMatchSnapshot()
})

test('FormLegend with additional class name', () => {
  const view = shallow(<FormLegend className="first-class">I am Legendary</FormLegend>)
  expect(view).toMatchSnapshot()
})

test('FormLegend with additional class name and props', () => {
  const view = shallow(
    <FormLegend className="first-class" id="legendary" data-label-name="Legendary">
      I am Legendary
    </FormLegend>
  )
  expect(view).toMatchSnapshot()
})
