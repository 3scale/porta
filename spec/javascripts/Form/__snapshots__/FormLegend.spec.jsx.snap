// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormLegend 1`] = `
<legend
  className="pf-c-form__legend "
>
  I am Legendary
</legend>
`;

exports[`FormLegend with additional class name 1`] = `
<legend
  className="pf-c-form__legend first-class"
>
  I am Legendary
</legend>
`;

exports[`FormLegend with additional class name and props 1`] = `
<legend
  className="pf-c-form__legend first-class"
  data-label-name="Legendary"
  id="legendary"
>
  I am Legendary
</legend>
`;
